src/base/bittorrent/session.cpp

```diff
@@ -2997,6 +3082,16 @@ void Session::handleTorrentCategoryChanged(TorrentHandle *const torrent, const Q
     emit torrentCategoryChanged(torrent, oldCategory);
 }
 
+void Session::handleTorrentTagAdded(const QString &tag, int new_tag_count)
+{
+    emit torrentTagAdded(tag, new_tag_count);
```

Comment:
The tag filter has a special "Untagged" filter for torrents that have 0 tags assigned to them. Any time you add a tag to a torrent that was previously Untagged, you need to remove it from the that filter. Any time you remove the last tag from a torrent, you need to add it to that filter. This is why a count of the number of tags is passed with the signal.